rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -479,7 +534,27 @@ impl CanisterQueues {
                     .expect("pop() should return the sender peeked above");
                 continue;
             };
-            let msg = pop_and_advance(input_queue, &mut self.pool);
+
+            let msg = if let Some(reference) = input_queue.pop() {
+                // Message must be either pooled; or a previously shed inbound response.
```

Comment:
We know that it must be either here or there because of the "No stale message in front of queue" invariant, right?

Reply:
Precisely.

An empty queue would have been skipped above (`else { ...; continue; }`). And a non-empty queue must have a non-stale message in front.